### ISSUE TITLE
Multiple books on the same host contaminate each other

### ISSUE THREAD

**Issue by Michael-F-Bryan** [azerupi/mdBook#455]
If multiple books are on the same host, changing the theme on one can magically change the theme on all the others. 

Steps to Reproduce:

1. Open up the first and second versions of the Rust Book in separate tabs
2. Go to the first book and change its theme
3. Refresh the second book's tab

You'll probably also see this if you use mdbook on multiple projects and have them hosted on GitHub Pages.

This is somewhat related to rust-lang/rust#44704 (see @steveklabnik's [comment]) in that local storage keys aren't handled as well as they could be. 

> mdbook stores the current theme in localstorage. This did change in newer/older mdbook, but it looks like this, basically:
>
> ```javascript
>    var theme = localStorage.getItem('theme');
>    if (theme === null) { theme = 'light'; }
> ```

Ideally instead of using a key like "theme", it'd be namespaced. So in the case of the first Rust Book you might use "/book/first-edition:theme".

There may also be other cases where local storage handling allows cross-book contamination.

[comment]: https://github.com/rust-lang/rust/issues/44704#issuecomment-330648133

**Comment by projektir** [comment]
> If multiple books are on the same host, changing the theme on one can magically change the theme on all the others.

Bug or feature? :grinning:

**Comment by azerupi** [comment]
Indeed, for some this can be a feature and for others this can be a bug. I am not sure what behavior we should adopt. If we conclude this is a bug, then prefixing all keys with the book's title should be enough.

**Comment by Michael-F-Bryan** [comment]
It might be worth asking someone on the Rust docs team, seeing as they're the group most affected by this. 

From a purity perspective I'd call this a bug. It's pretty much a web version of global mutable state and data races.

**Comment by budziq** [comment]
It might be nice for people wanting to set coherent dark theme everywhere but its unexpectedly shared between domains. Moreover, not all mdBook clients are related to docs team or rust at all. So certainly a bug!

**Comment by azerupi** [comment]
>shared between domains

I doubt that, localstorage, or any other web storage, is scoped per domain as far as I know. I am not sure how sub-domains are handled though.

**Comment by budziq** [comment]
Right I forgot about it!

**Comment by Michael-F-Bryan** [comment]
It doesn't just affect dev environments though. For example, I've got multiple repositories which use mdbook and are published to GitHub Pages (e.g. my [FFI guide]). These will all be affected, likewise anyone else who has multiple books published to GitHub pages will be the same.

And of course, other big case of multiple books on the same domain is the multiple Rust Book versions and the Nomicon.

I agree that books contaminating each others themes is quite a minor issue. Although if it's happened once, chances are other places using local storage will have the same issue.

[FFI guide]: https://michael-f-bryan.github.io/rust-ffi-guide/
